### ISSUE TITLE
Possible memory leak

### ISSUE THREAD

**Issue by ScruffR** [ubidots/ubidots-particle#32]
While assisting a Particle forum member in [this thread](https://community.particle.io/t/photon-stops-responding-randomly/37115/18?u=scruffr), I think I stumbled over a possible memory leak in this library.

Not all code paths in this function seem to free the 700byte chunk allocated by `sendAll()` and no other function seems to clean this up later on either.

```
bool Ubidots::sendAllTCP(char* buffer) {
    int i = 0;
    while (!_client.connected() && i < 6) {
        i++;
        if (_debug) {
            Serial.println("not connected, trying to connect again");
        }
        _client.connect(_server, PORT);
        if (i == 5) {
            if (_debug) {
                Serial.println("Max attempts to connect reached, data could not be sent");
            }
            return false;
        }
    }
    if (_client.connected()) {        // Connect to the server
        if (_debug) {
            Serial.println("Sending data");
        }
        _client.println(buffer);
        _client.flush();
        _client.stop();
        free(buffer);
        _currentValue = 0;
        return true;
    }
    _currentValue = 0;
    return false; // If any of the above conditions is reached, returns false to indicate an unexpected issue
}
```

**Comment by jotathebest** [comment]
Hi there, not all the paths free the buffer, my assumption when I implemented the library was to avoid to lose data, if the device cannot connect to the cloud after a max attempts, the user will not want to lose the data already read and stored inside the buffer. BTW, I believe that this may be a wrong assumption made by me as there is another ways to stored data if there is not connection (i.e a SD), so the best in this case is to add a method, freeBuffer(), that frees all the memory used by the library.

**Comment by ScruffR** [comment]
You may not lose the data but I don't see any attempt in your code to pick that "preserved" data up for later upload. Consequently I don't see any benefit in keeping the allocated packet on the heap. 
If such a "pick-up" logic would be implemented it might make sense, but the premisse has to be: "Once allocated memory *must* be freed eventually"

Alternatively you could opt for keeping the "raw" data in the individual "variables" till successfully published and keep reallocating packets on the heap but freeing them even on failing publishes.

**Comment by jotathebest** [comment]
Fixed on release 2.1.13
